firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapDecoder.kt

```diff
@@ -0,0 +1,244 @@
+     * decoded as the field value of the [Serializable] object.
+     *
+     * Note: An [IllegalArgumentException] will be thrown if the [elementIndex] is pointing to a
+     * null value from [decodeValueList].
+     */
+    private fun validateDecodedElementNotNull(): Any =
+        decodeValueList.elementAt(elementIndex - 1)
+            ?: throw IllegalArgumentException(
+                "Got a null value while trying to decode a not null field."
+            )
+
+    final override val serializersModule: SerializersModule = EmptySerializersModule
+
+    final override fun beginStructure(descriptor: SerialDescriptor): CompositeDecoder {
+        val innerCompositeObject = getCompositeObject(elementIndex)
+        when (descriptor.kind) {
+            is StructureKind.CLASS -> {
+                val innerMap = innerCompositeObject as Map<String, Any>
+                return FirestoreMapDecoder(innerMap, docRef)
+            }
+            is StructureKind.LIST -> {
+                val innerList = innerCompositeObject as List<Any>
+                return FirestoreListDecoder(innerList, docRef)
+            }
+            else -> {
+                throw Exception("Incorrect format of nested data provided: <$innerCompositeObject>")
+            }
+        }
+    }
+
+    /**
+     * Returns the nested structured object that needs to be decoded. The nested structured object
+     * could be either a nested [Serializable] object or a nested List.
+     */
+    private fun getCompositeObject(elementIndex: Int): Any? {
+        return when (elementIndex) {
+            START_INDEX ->
+                nestedObject // the custom object itself is the first structured obj to decode
+            else -> decodeValueList[elementIndex - 1]
+        }
+    }
+}
+
+/**
+ * The entry point of Firestore Kotlin deserialization process. It decodes a nested map of Firestore
+ * supported types to a [Serializable] Kotlin object.
+ *
+ * For a [Serializable] object, at compile time, a serializer will be generated by the Kotlin
+ * serialization compiler plugin (or a custom serializer can be manually passed in). The structure
+ * information of the [Serializable] object will be recorded inside of the serializer’s descriptor
+ * (i.e. the name/type of each property to be encoded, the annotations on each property).
+ *
+ * Based on the descriptor’s information, during the decoding process, a nested map will be send to
+ * the decoder. Decoder will loop through all the key-value pairs in the nested map, use the key to
+ * find the target field from the [SerialDescriptor], then assign the value to the field.
+ *
+ * @param nestedObject The nested map that that needs to be decoded to a @[Serializable] object.
+ * @param docRef The [DocumentReference] where this nested map is obtained from.
+ */
+private class FirestoreMapDecoder(nestedObject: Map<String, Any?>, docRef: DocumentReference) :
```

Comment:
nestedobject is actually a map.

Reply:
this parameter I changed the name to `val nestedMap`, the same parameter for list decoder, I changed to `override val decodeValueList`